client/modules/wj-util/src/index.ts

```diff
@@ -156,3 +156,138 @@ export function perfy(meta?: string, threshold?: number): () => number {
+  let timeout: any
+  return function (this: any, ...args: Parameters<T>) {
+    clearTimeout(timeout)
+    timeout = setTimeout(() => void fn.call(this, ...args), wait)
+  }
+}
+
+/** Waits until the specified function returns `true`.
+ *  It will call the specified async function to determine the polling interval.
+ *  If none is given, it will poll every 100ms. */
+export async function waitFor(
+  conditionFn: () => Promisable<boolean>,
+  asyncTimerFn: () => Promise<void> = () => sleep(100)
+) {
+  while ((await conditionFn()) === false) {
+    await asyncTimerFn()
+    continue
+  }
+  return true
+}
+
+/** Returns a new 'locked' async function, constructed using the specified function.
+ *  A locked asynchronous function will only allow a singular instance of itself
+ *  to be running at one time.
+ *  Additional calls will cause the function to wait until they can be ran. */
+export function createLock<T extends AnyFunction>(fn: T) {
+  let locked: boolean
+  return async (...args: Parameters<T>) => {
+    if (locked) await waitFor(() => locked === false)
+    locked = true
+    const result = (await fn(...(args as any))) as PromiseValue<T>
+    locked = false
+    return result
+  }
+}
+
+/** Returns a function that will be "queued" to execute only on animation frames.
+ *  Calling multiple times will run only once on the next requestAnimationFrame.
+ *  @example
+ *  const coolFunc = createAnimQueued(function niceFunc(args) => { 'stuff' })
+ *  coolFunc()
+ *  coolFunc() // doesn't run as the previous call is already queued */
+export function createAnimQueued<T extends AnyFunction>(fn: T) {
+  let queued: boolean
+  return (...args: Parameters<T>): void => {
+    if (queued !== true) {
+      queued = true
+      requestAnimationFrame(async () => {
+        await fn(...(args as any))
+        queued = false
+      })
+    }
+  }
+}
+
+const HAS_IDLE_CALLBACK = "requestIdleCallback" in window
+
+/** Safely calls `requestIdleCallback` in an awaitable `Promise`. */
+export function idleCallback<T extends AnyFunction<any>>(
+  cb: T,
+  timeout = 100
+): Promise<ReturnType<T>> {
+  if (!HAS_IDLE_CALLBACK) {
+    return new Promise(resolve => setTimeout(() => resolve(cb()), timeout))
+  } else {
+    return new Promise(resolve =>
+      (window as any).requestIdleCallback(() => resolve(cb()), { timeout })
+    )
+  }
+}
+
+/** See `createAnimQueued` for a description of how this function works.
+ *  The only difference is that this function uses `requestIdleCallback` instead.
+ *  If `requestIdleCallback` isn't available, it will use `createAnimQueued` instead.
+ * @see createAnimQueued */
+export function createIdleQueued<T extends AnyFunction>(fn: T, timeout = 100) {
+  if (!HAS_IDLE_CALLBACK) return createAnimQueued(fn)
+  let queued: boolean
+  return (...args: Parameters<T>): void => {
+    if (queued !== true) {
+      queued = true
+      ;(window as any).requestIdleCallback(
```

Comment:
Is this something prettier did? It looks very weird.

Reply:
Yeah, sometimes code gets mangled, especially when you're doing a weird type cast like this because there isn't a decent way in TS to assert the existence of something. There is proposals for it.

Reply:
I was referring to the `;`, I understand why `(x as any)` has to happen on occasion.

Reply:
Oh, sorry, yeah - that's Prettier. The idea is, when you're running without semicolons, is that you put the semicolon at the _start_ of the line when you need them, because it's super obvious.

A way of avoiding that would be TypeScript's original type cast assertion - `<any>window`. However, it's somewhat discouraged because it's incompatible with JSX/TSX.